lib/pbench/cli/server/shell.py

```diff
@@ -83,21 +83,27 @@ def keycloak_connection(
     # which defaults to 120.0s More detailed documentation on Retry and
     # backoff_factor can be found at:
     # https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry
-    try:
-        retry = Retry(total=8, backoff_factor=2)
-        adapter = HTTPAdapter(max_retries=retry)
-        session.mount("http://", adapter)
-        response = session.get(f"{oidc_server}/health")
+    retry = Retry(total=8, backoff_factor=2)
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount("http://", adapter)
+
+    # We will also need to retry the connection if the health status is not UP
+    for i in range(5):
+        try:
+            response = session.get(f"{oidc_server}/health")
+        except Exception:
+            logger.exception("Error connecting to the OIDC client")
+            return return_val + i
         if response.status_code == HTTPStatus.OK and response.json()["status"] == "UP":
```

Comment:
I believe that you can configure the `Retry` to reissue the request on basically anything other than `OK`, which would allow you to simplify this check; and, it would engage the backoff automatically for those cases.

Reply:
I considered that. I could do something like
```
status_forcelist = tuple(x for x in requests.status_codes._codes if x != 200)
```
since the Retry object does not have a provision to specify a retry on `all_except_these_status_codes`. However, now I also wonder if there is any point in retrying on 401/403 since any number of retries will not resolve that unless we modify the request to include the required headers (if we ever need to). 
Specifying the `status_forcelist` as above means I can also get rid of sleep because the request can automatically get retried.

Reply:
Yeah, we (you?) would need to dive into the possible result values, but surely they divide into three buckets:  one(s) that we would consider "success"; ones which we would consider fatal errors; and ones that we would want to retry.  I suppose it would make the most sense to set up the `Retry` object with only the ones that we would want to retry (perhaps there's not that many, and perhaps the default value covers them...as I said, someone would have to think about that 🙂), and then when we get a response back which doesn't get caught by the `Retry` mechanism, we either return immediately with a failure or we look to see if it's reporting the server state.  At that point, we only engage the outer loop if we got a "success" and a not-"UP".

However, I think you can configure the `Retry` object to help, there, too.  In addition to setting the response codes which trigger a retry, I think you can have other responses reported as exceptions.  So, you can probably arrange it so that the _only_ normal return from `get()` is actually a `200`, and everything else is either retried or raised as an exception.